src/db.rs

```diff
@@ -260,27 +262,60 @@ impl DbInner {
         }
     }
 
-    async fn replay_wal(&self) -> Result<(), SlateDBError> {
+    async fn replay_wal(self: &Arc<Self>) -> Result<(), SlateDBError> {
         let wal_id_last_compacted = self.state.read().state().core.last_compacted_wal_sst_id;
         let wal_sst_list = self
             .table_store
             .get_wal_sst_list(wal_id_last_compacted)
             .await?;
         let mut last_sst_id = wal_id_last_compacted;
+        let mut wal_tasks = Vec::new();
+        // we use a BTreeMap here to preserve order of sst_id
+        let mut wal_replay_buffers = BTreeMap::new();
+        let mut wal_replay_errors = Vec::new();
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(
+            self.options.maximum_parallel_tasks_limit,
+        ));
+
+        // constructs a list of tokio tasks to asynchronously fetch wal sst
         for sst_id in wal_sst_list {
-            last_sst_id = sst_id;
-            let sst = self.table_store.open_sst(&SsTableId::Wal(sst_id)).await?;
-            let sst_id = match &sst.id {
-                SsTableId::Wal(id) => *id,
-                SsTableId::Compacted(_) => return Err(SlateDBError::InvalidDBState),
-            };
-            let mut iter = SstIterator::new(&sst, self.table_store.clone(), 1, 1, true).await?;
-            // iterate over the WAL SSTs in reverse order to ensure we recover in write-order
-            // buffer the WAL entries to bulk replay them into the memtable.
-            let mut wal_replay_buf = Vec::new();
-            while let Some(kv) = iter.next_entry().await? {
-                wal_replay_buf.push(kv);
+            let this = self.clone();
+            let semaphore = semaphore.clone();
+            wal_tasks.push(tokio::spawn(async move {
+                match semaphore.acquire().await {
+                    Ok(_) => this.fetch_wal(sst_id).await,
+                    Err(e) => Err(SlateDBError::AcquireError(e)),
+                }
+            }));
+        }
+
+        // executes all the `fetch_wal` tasks in parallel
+        let results = join_all(wal_tasks).await;
+
+        // collect the results in order of sst_id
+        for result in results {
+            match result {
+                Ok(res) => match res {
+                    Ok(r) => {
+                        // maps `sst_id` to `wal_replay_buffer`
+                        // regardless of the insertion order BTreeMap is sorted by key
+                        wal_replay_buffers.insert(r.0, r.1);
+                    }
+                    Err(e) => wal_replay_errors.push(e),
+                },
+                Err(e) => {
```

Comment:
We'd only get a join error if the task panic'd, so I think it's reasonable to just call `expect` (which panics when there is an error) on the `Result` returned from `join_all`. This would behave almost the same as if the wal fetch was not done concurrently. This is also a pretty meaningless error for the user.